Fix corner case uses of memory_buffer_alloc.c

andresag01 opened the pull request:
This change fixes the following corner cases in memory_buffer_alloc.c:
* Calling `verify_chain()` on an uninitialized chain. Would previously cause a `NULL` dereference.
* Allocating a memory block large enough to overflow the length variable.
* Initializing the allocator with a buffer smaller than the size of a memory header.

This PR is based on the information recorded in https://github.com/ARMmbed/mbedtls/issues/639.

**NOTES:**
* This PR includes tests.
* The changes need to be backported to 2.7 (use this PR) and 2.1 (#1009)

andresag01 commented on library/memory_buffer_alloc.c:
```diff
@@ -247,6 +249,8 @@ static void *buffer_alloc_calloc( size_t n, size_t size )
 
     if( n != 0 && len / n != size )
```
In https://github.com/ARMmbed/mbedtls/issues/639 it was proposed that if the size of the requested allocation is `0` then the allocator should return `NULL`. The C99 standard states the following:
```
If the  size  of  the  space  requested  is  zero,  the  behavior  is  implementation-
defined:  either  a  null  pointer  is  returned,  or  the  behavior  is  as  if  the  size  were  some
nonzero value, except that the returned pointer shall not be used to access an object.
```
The problem is that some parts of the library assume that allocating 0 will return a non-NULL value on success. Making this change causes test failures, so the behaviour is not changed as it is compliant with the C standard.

RonEld replied:
@andresag01 In my opinion, we should treat 0 size allocation as NULL, and fix the code where it is treated different

sbutcher-arm replied:
@andresag01 - I read the bug differently. That's because `calloc()` can return `NULL` or non-`NULL` values, we need to be able to handle both because we can't always depend on the underlying implementation to do what we expect. If, as you say, "some parts of the library assume that allocating 0 will return a non-NULL value", surely we need to fix this to accommodate both types of behaviour?

@andresag01 - I think you'd best raise a new bug to address other parts of the library.

I also think we need to treat a size 0 allocation as NULL.

andresag01 commented:
@RonEld, @sbutcher-arm: I have modified the allocator to return NULL when  allocating 0 bytes. And fixed two places in the library (pem.c and x509_crl.c) that worked under the assumption that allocating 0 returns a non-NULL value.

andresag01 commented:
@RonEld: Could you take a look a this again?

RonEld commented on library/pem.c:
```diff
@@ -410,7 +410,7 @@ int mbedtls_pem_write_buffer( const char *header, const char *footer,
     }
 
-    if( ( encode_buf = mbedtls_calloc( 1, use_len ) ) == NULL )
+    if( use_len != 0 && ( encode_buf = mbedtls_calloc( 1, use_len ) ) == NULL )
```
to make it more readable, please surround the statement `( encode_buf = mbedtls_calloc( 1, use_len ) ) == NULL` with parentheses

RonEld commented on library/x509_crl.c:
```diff
@@ -257,7 +257,7 @@ int mbedtls_x509_crl_parse_der( mbedtls_x509_crl *chain,
     int ret;
     size_t len;
-    unsigned char *p, *end;
+    unsigned char *p = NULL, *end;
```
I would also initialize `end`

RonEld commented on library/x509_crl.c:
```diff
@@ -294,7 +294,7 @@ int mbedtls_x509_crl_parse_der( mbedtls_x509_crl *chain,
      * Copy raw DER-encoded CRL
      */
-    if( ( p = mbedtls_calloc( 1, buflen ) ) == NULL )
+    if( buflen != 0 && ( p = mbedtls_calloc( 1, buflen ) ) == NULL )
```
to make it more readable, please surround the statement `( p = mbedtls_calloc( 1, buflen ) ) == NULL` with parentheses

RonEld commented on library/memory_buffer_alloc.c:
```diff
@@ -581,20 +584,24 @@ void mbedtls_memory_buffer_alloc_init( unsigned char *buf, size_t len )
 #endif
 
-    if( (size_t) buf % MBEDTLS_MEMORY_ALIGN_MULTIPLE )
+    if( len < sizeof( memory_header ) + MBEDTLS_MEMORY_ALIGN_MULTIPLE )
```
Is this the correct behavior? This function doesn't have a return code, but eventually, this is an error case. A user calling this function doesn't know that the input is incorrect

andresag01 replied:
I think this is the correct behaviour. The buffer provided to memory manage must be at least as high as the memory header. If this is not the case, the internal `buffer_alloc_ctx` will be zeroized before returning and further calls to, say `mbedtls_calloc()` will signal `allocation failed` by returning a `NULL` pointer.

RonEld replied:
ok thanks, I think you mean `buffer_alloc_calloc()` though

andresag01 commented:
@RonEld, @sbutcher-arm: I have reworked the PR to address comments. Please review.

gilles-peskine-arm commented:
CI failed only due to timing tests (known issue), ok to merge.

gilles-peskine-arm commented:
Failure in `x509parse-suite` (https://jenkins-internal.mbed.com/job/mbedtls-pr/424/execution/node/205/log/). **DO NOT MERGE** pending investigation.

gilles-peskine-arm commented on library/x509_crl.c:
```diff
@@ -294,7 +294,7 @@ int mbedtls_x509_crl_parse_der( mbedtls_x509_crl *chain,
      * Copy raw DER-encoded CRL
      */
-    if( ( p = mbedtls_calloc( 1, buflen ) ) == NULL )
+    if( buflen != 0 && ( ( p = mbedtls_calloc( 1, buflen ) ) == NULL ) )
```
The check for `buflen == 0` should be made earlier. See https://github.com/ARMmbed/mbedtls/pull/1009/files#r153007520

andresag01 commented:
@gilles-peskine-arm: I have fixed the issue.

gilles-peskine-arm commented on library/x509_crl.c:
```diff
@@ -294,7 +294,9 @@ int mbedtls_x509_crl_parse_der( mbedtls_x509_crl *chain,
-    if( buflen != 0 && ( ( p = mbedtls_calloc( 1, buflen ) ) == NULL ) )
+    if( buflen == 0 )
+        return( MBEDTLS_ERR_X509_INVALID_FORMAT );
+    else if( ( p = mbedtls_calloc( 1, buflen ) ) == NULL )
```
It doesn't change the behavior, but I'd prefer not to put an `else` here, because it's misleading for readers: if `buflen==0`, the code below the `else` is no executed either. And since the result of the allocation is stored in a variable, there's no call for putting the assignment inside the `if`, so I'd prefer

```
if( buflen == 0 )
    return( MBEDTLS_ERR_X509_INVALID_FORMAT );
p = mbedtls_calloc( 1, buflen );
if( ( p == NULL )
    return( MBEDTLS_ERR_X509_ALLOC_FAILED );
```

andresag01 replied:
Ok, I think this is a really minor thing about coding style. I prefer the way I originally wrote it, but I changed the code to suit your preferred style in the latest commit.

gilles-peskine-arm commented:
CI failed only due to timing tests (known issue), ok to merge.

andresag01 commented:
@gilles-peskine-arm: Thanks for letting me know. The problem is yet another piece of code in the library attempting to allocate a 0-length buffer. I fixed it and will push the changes soon. I have also rebased the PR and run all.sh locally before passing it back to you.

andresag01 commented:
@gilles-peskine-arm: I have rebased on top of development and fixed the issue you raised. I will now update the backports

andresag01 commented:
@gilles-peskine-arm: Thanks for the review. I will do the backport to 2.7 as soon as I can.

gilles-peskine-arm commented:
@andresag01 Since this PR forked before 2.7.0, you don't need to make a 2.7 backport: the same commit can be merged into 2.7 and development. We'll merge this PR into both `mbedtls-2.7` and `development`.

andresag01 commented:
@gilles-peskine-arm: Alright thanks. I will not do the backport then.
